helix-core/src/test/java/org/apache/helix/integration/task/TestPurgeJobWithoutConfig.java

```diff
@@ -0,0 +1,110 @@
+import org.testng.annotations.Test;
+
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * This test checks whether a job without config is being removed from the DAG.
+ */
+public class TestPurgeJobWithoutConfig extends TaskTestBase {
+
+  @BeforeClass
+  public void beforeClass() throws Exception {
+    super.beforeClass();
+  }
+
+  @Test
+  public void testPurgeJobWithoutConfig() throws Exception {
+    // Timeout per task has been set to be a large number.
+    final long timeout = 60000L;
+    String jobQueueName = TestHelper.getTestMethodName();
+
+    JobQueue.Builder jobQueue = TaskTestUtil.buildJobQueue(jobQueueName);
+
+    JobConfig.Builder jobBuilder0 = JobConfig.Builder.fromMap(WorkflowGenerator.DEFAULT_JOB_CONFIG)
+        .setTimeoutPerTask(timeout).setMaxAttemptsPerTask(1)
+        .setJobCommandConfigMap(ImmutableMap.of(MockTask.JOB_DELAY, "1000"));
+
+    JobConfig.Builder jobBuilder1 = JobConfig.Builder.fromMap(WorkflowGenerator.DEFAULT_JOB_CONFIG)
+        .setTimeoutPerTask(timeout).setMaxAttemptsPerTask(1)
+        .setJobCommandConfigMap(ImmutableMap.of(MockTask.JOB_DELAY, "1000"));
+
+    JobConfig.Builder jobBuilder2 = JobConfig.Builder.fromMap(WorkflowGenerator.DEFAULT_JOB_CONFIG)
+        .setTimeoutPerTask(timeout).setMaxAttemptsPerTask(1)
+        .setJobCommandConfigMap(ImmutableMap.of(MockTask.JOB_DELAY, "10000"));
+
+    jobQueue.enqueueJob("JOB0", jobBuilder0);
+    jobQueue.enqueueJob("JOB1", jobBuilder1);
+    jobQueue.enqueueJob("JOB2", jobBuilder2);
+
+    _driver.start(jobQueue.build());
+
+    // Wait until Queue is running.
+    _driver.pollForWorkflowState(jobQueueName, TaskState.IN_PROGRESS);
+
+    // Check the JOB1 is completed
+    String nameSpacedJobName = TaskUtil.getNamespacedJobName(jobQueueName, "JOB1");
+    _driver.pollForJobState(jobQueueName, nameSpacedJobName, TaskState.COMPLETED);
+
+    // Remove the config associated with JOB1
+    boolean configDeleted =
+        removeWorkflowJobConfig(_manager.getHelixDataAccessor(), nameSpacedJobName);
+    Assert.assertTrue(configDeleted);
+
+    // Check whether JOB1 has been successfully removed from the DAG
+    boolean haveJobDeletedFromDag = TestHelper.verify(() -> {
+      WorkflowConfig workflowConfig = _driver.getWorkflowConfig(jobQueueName);
+      JobDag dag = workflowConfig.getJobDag();
+      return !dag.getAllNodes().contains(nameSpacedJobName);
+    }, 60 * 1000);
+    Assert.assertTrue(haveJobDeletedFromDag);
+  }
+
+  /**
+   * This function removes the job config from the ZooKeeper
+   * @param accessor
+   * @param workflowJobResource
+   * @return
+   */
+  private static boolean removeWorkflowJobConfig(HelixDataAccessor accessor,
+      String workflowJobResource) {
+    PropertyKey cfgKey = accessor.keyBuilder().resourceConfig(workflowJobResource);
+    if (accessor.getPropertyStat(cfgKey) != null) {
+      if (!accessor.removeProperty(cfgKey)) {
+        return false;
+      }
+    }
+    return true;
+  }
```

Comment:
Isn't this duplicate code? Did you copy & paste this here?

Reply:
Fixed.